### ISSUE TITLE
CPU feature: SHA

### ISSUE THREAD

**Issue by cperciva** [Tarsnap/libcperciva#44]
We should detect the Intel SHA extensions and use them to speed up alg/sha256 when available.

**Comment by noloader** [comment]
The following is a partial patch to support SHA256 on Intel platforms. I don't know the best way to integrate it into scrypt, so other will have to finish it.

SHA256 on Intel platforms will run about 3.8 cycles per byte.

You can test the code on an Intel Goldmont or Goldmont+ machine, like a [Celeron J3455](https://www.amazon.com/dp/B01LYCDG4H). If you have a [GCC Compile Farm account](https://cfarm.tetaneutral.net/users/new/) then gcc68, which is the the Ryzen 7 1700x, has SHA extensions.

Take as much or as little as you want from it.

Here is the patch: [sha256.diff.zip](https://github.com/Tarsnap/libcperciva/files/1866110/sha256.diff.zip)

**Comment by gperciva** [comment]
Hi @noloader,

I'm taking a look at integrating it into `sha256.c`.  In the mean-time, see https://github.com/Tarsnap/libcperciva/commit/7a3113828fd277b10a8c475ccd52e79d9bab50e6 for info (and fix) for the clang compile failure.

I'm working on the branch https://github.com/Tarsnap/libcperciva/tree/sha256-shani

**Comment by noloader** [comment]
Good job @cperciva. You make this stuff look so easy. I'm quite envious.

Yeah, this compile failure is very unusual to me. I've been using the SHA intrinsics since about November 2016. I've never encountered the warning or error.

I do a lot with inline assembly and intrinsics. Maybe this is a Clang hiccup. Clang gives trouble in some places one would not expect, like failing to generate a negate in inline assembly and failing and complaining an address was unaligned when the data did not have alignment requirements.

This question may help if it gets answered: [Problems with hexadecimal constant, _mm_set_epi64x and sign conversion](https://lists.llvm.org/pipermail/cfe-users/2018-March/001284.html).

**Comment by gperciva** [comment]
Hi @noloader,

I don't think it's a clang "hiccup".  It recognizes that `0xA4506CEB90BEFFFA` (for example) cannot fit into a signed long long int.  If you interpret that string as a number, it simply cannot fit.

The problem is that `_mm_set_epi64x()` wants signed integers.  So if we want to use that function, then either we need to express `0xA4506CEB90BEFFFA` as a negative value, or tell the compiler to change the unsigned value into a signed value.

Expressing `0xA4506CEB90BEFFFA` as a negative value would be silly, because the SHA256 docs use unsigned constants, so turning some of them into negative values would make it really hard to double-check the code.  So the remaining option is "what's the best way to tell the compiler to change `0xA...` into a signed value?".

(alternatively, we could use `_mm_set_epi64()`, which takes `__m64` values.  I'm not certain if that's easier than adding a cast or macro, but it's something else I'm looking into)

**Comment by gperciva** [comment]
Never mind that last comment -- it looks like `__m64` is a signed value.  Because of *course* we all love dealing with 63 bits of data, plus 1 bit of signed.

**Comment by cperciva** [comment]
Well that's an interesting bug in the standard.  INT64_C(foo) produces a constant of *unsigned* type if foo is greater than 2^63-1.

How about something like
```
#define I64(x) ((UINT64_C(x) >= UINT64_C(0x8000000000000000)) ? -(int64_t)(UINT64_C(0xffffffffffffffff) - UINT64_C(x))-1 : INT64_C(x))
```
?

**Comment by cperciva** [comment]
Simply casting it isn't standards-compliant, since casting an out-of-range value to a signed type yields undefined (overflow) behaviour.

**Comment by gperciva** [comment]
`I64()` as written above produces:
```
../../crypto/crypto_sha256_shani.c:231:7: error: operand of ? changes
      signedness: 'unsigned long' to 'long long' [-Werror,-Wsign-conversion]
                    I64(0xA4506CEB90BEFFFA)));
                    ^~~~~~~~~~~~~~~~~~~~~~~
../../crypto/crypto_sha256_shani.c:15:4: note: expanded from macro 'I64'
        : INT64_C(x))
          ^~~~~~~~~~
/usr/include/stdint.h:286:22: note: expanded from macro 'INT64_C'
#  define INT64_C(c)    c ## L
                        ^~~~~~
<scratch space>:34:1: note: expanded from here
0xA4506CEB90BEFFFAL
^~~~~~~~~~~~~~~~~~~
```

Casting the `:` portion makes it work.  That is to say, I'm currently using:
```
/* Get an (int64_t). */
#define I64(x) ((UINT64_C(x) >= UINT64_C(0x8000000000000000))           \
        ? -(int64_t)(UINT64_C(0xffffffffffffffff) - UINT64_C(x)) - 1    \
        : (int64_t)INT64_C(x))
```

I'm also thinking that it might be nice to use `_mm_set_epi32` instead of `_mm_set_epi64x`, since `sha256.c` already contains `static const uint32_t Krnd[64]`.  As an added bonus, all those magic numbers are in one place in `Krnd`, which is easier to check than a whole bunch of numbers spread throughout the code.

**Comment by noloader** [comment]
@cperciva,

> Simply casting it isn't standards-compliant, since casting an out-of-range value to a signed type yields undefined (overflow) behaviour.

Thanks. I've never given that much thought. I guess the Clang folks expect someone to `memcpy` it from an unsigned type to a signed type... derp.

**Comment by noloader** [comment]
@gperciva,

> I don't like to leave this in limbo... 

I can test this later tonight when I get home. I have access to the Compile Farm, but I don't have my SSH keys with me. I also have a Goldmont at the house.

**Comment by gperciva** [comment]
Thanks, emailed!  I don't have a branch ready for testing yet; I ran out of steam when I realized that I couldn't test any of my changes.  I should have it ready in a day or so.

**Comment by cperciva** [comment]
Apparently Intel makes an emulator which can simulate all of their different CPUs: https://software.intel.com/en-us/articles/intel-software-development-emulator

**Comment by noloader** [comment]
> Apparently Intel makes an emulator ...

Yes, I forgot about that. You can use the emulator, too.

If you want to add the emulator to Travis for CI testing, then see [SHA-Intrinsics | travis.yml](https://github.com/noloader/SHA-Intrinsics/blob/master/.travis.yml). @pavelkryukov added the process to the testing.

**Comment by pavelkryukov** [comment]
> You will need to compile with -msse4.2 -msha for the SHA extensions. SHA256 on Intel platforms will run about 3.8 cycles per byte.

I prefer an alternative syntax provided by GCC and Clang

```c
__attribute__ ((target("sse4.1,sha")))
void function_that_uses_sha_intrinsics( /* arguments */)
{

}
```

There are two advantages
1.  SSE code is not emitted for other functions define in the same translation unit
2. You don't have to change CMake, Makefile etc. to add command line options

> If you want to add the emulator to Travis for CI testing, then see [SHA-Intrinsics | travis.yml](https://github.com/noloader/SHA-Intrinsics/blob/master/.travis.yml).

The key point is to set SDE_URL and SDE_ARCHIVE hidden variables in Travis settings. They should point to Intel SDE distribution on your server, and they should be hidden as the license does not allow redistribution of the tool.

There is an alternative approach by https://github.com/marehr/intel-sde-downloader tool — it automatically accepts Intel license and downloads SDE directly from Intel site.

**Comment by cperciva** [comment]
Fixed by #140
